### PR TITLE
Define classes for a unified event system

### LINKED ISSUE #11: [Phase2][2 weeks]: Proposal for New unified logging backend: Event storage and writer
We want to take the following steps towards creating a unified backend API. We will do the task in several small steps that come together in the end.

**Anticipated timeline**: 2 weeks for this task

# High level design

Step 1: Create a  storage and writer definition

Create `vissl/utils/events.py`

```
# https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/history_buffer.py
from fvcore.common.history_buffer import HistoryBuffer

_VISSL_EVENT_STORAGE_STACK = []


def get_event_storage():
    return _VISSL_EVENT_STORAGE_STACK[-1]
   

def create_event_storage()
    _VISSL_EVENT_STORAGE_STACK.append(VisslEventStorage())
    

    
class VisslEventWriter:
    """
    Base class for writers that obtain events from :class:`VisslEventStorage` 
    and process them.
    """

    def write(self):
        raise NotImplementedError

    def close(self):
        pass
  
  
class VisslEventStorage:
    """
    The user-facing class that stores the running metrics
    and all the logging data in one place. The Storage can
    be updated anywhere in the training step. The storage
    is used by several writers/hooks to write the data to
    several backends (json, tensorboard, etc)
    """

    def __init__(self, start_iter=0):
        """
        Args:
            start_iter (int): the iteration number to start with
        """
        self._history = defaultdict(HistoryBuffer)
        self._latest_scalars = {}
        self._iter = start_iter
        # self._vis_data = []     # later for tensorboard
        # self._histograms = []   # later for tensorboard

    def put_scalar(self, name, value):
        """
        Add a scalar `value` to the `HistoryBuffer` associated with `name`.
        """
        history = self._history[name]
        value = float(value)
        history.update(value, self._iter)
        self._latest_scalars[name] = (value, self._iter)

    def put_scalars(self, **kwargs):
        """
        Put multiple scalars from keyword arguments.
        Examples:
            storage.put_scalars(loss=my_loss, accuracy=my_accuracy)
        """
        for k, v in kwargs.items():
            self.put_scalar(k, v)

    def history(self, name):
        """
        Returns:
            HistoryBuffer: the scalar history for name
        """
        ret = self._history.get(name, None)
        if ret is None:
            raise KeyError(f"No history metric available for {name}!")
        return ret

    def histories(self):
        """
        Returns:
            dict[name -> HistoryBuffer]: the HistoryBuffer for all scalars
        """
        return self._history

    def latest(self):
        """
        Returns:
            dict[str -> (float, int)]: mapping from the name of each scalar to the most
                recent value and the iteration number its added.
        """
        return self._latest_scalars

    @property
    def iter(self):
        """
        Returns:
            int: The current iteration number.
        """
        return self._iter

    @iter.setter
    def iter(self, val):
        self._iter = int(val)

    def clear_images(self):
        return NotImplementedError

    def clear_histograms(self):
        return NotImplementedError
    
    def put_histogram(self, hist_name, hist_tensor, bins=1000):
        return NotImplementedError
     
    def put_image(self, img_name, img_tensor):
        # implement later for tensorboard
        return NotImplementedError
```

Step 2: Init the storage in VISSL trainer
```
# In the VISSL trainer, we will initiate a storage object that will be updated over
# time as the training progresses
# https://github.com/facebookresearch/vissl/blob/master/vissl/trainer/trainer_main.py#L89

class SelfSupervisionTrainer(object):
    def __init__(
        self,
        cfg: AttrDict,
        dist_run_id: str,
        checkpoint_path: str = None,
        checkpoint_folder: str = None,
        hooks: List[ClassyHook] = None,
    ):
        .....
        self.local_rank, self.distributed_rank = get_machine_local_and_dist_rank()
        self.task.initiate_vissl_event_storage()
        
# Create the initiaite_vissl_event_storage() in train_task.py
# https://github.com/facebookresearch/vissl/blob/master/vissl/trainer/train_task.py
@register_task("self_supervision_task")
class SelfSupervisionTask(ClassificationTask):
    def __init__(self, config: AttrDict):
        super().__init__()
        .....
        self._event_storage = None
        ....
    
    def initiate_vissl_event_storage(self):
        from vissl.utils.events import create_event_storage, get_event_storage
        
        create_event_storage()
        self._event_storage = get_event_storage()
    
     @property
    def event_storage(self):
        return self._event_storage
```

Step3:  Now the next step for us is to use the storage created and update it during the training
https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L152-L159 and https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L166-L180 in this steps, everything should be logged to the storage 
```
task.event_storage.put_scalars(...)
```

Step 4:  We will now write a writer that will take the storage and hooks will use the writer instead
```
# in vissl/utils/events.py
class JsonWriter(VisslEventWriter):
 
  def __init__(self, json_file):
     """
     Args:
         json_file: path to the json file. New data will be appended if the file 
                    exists.
     """
     self._file_handle = PathManager.open(json_file, "a")
  
  def write(self):
      storage = get_event_storage()
      to_save = defaultdict(dict)

      for k, (v, iter) in storage.latest.items():
          # keep scalars that have not been written
          to_save[iter][k] = v
      if len(to_save):
          all_iters = sorted(to_save.keys())

      for itr, scalars_per_iter in to_save.items():
          scalars_per_iter["iteration"] = itr
          self._file_handle.write(json.dumps(scalars_per_iter, sort_keys=True) + "\n")
      self._file_handle.flush()

  def close(self):
      self._file_handle.close()
      

# initialize the writer in trainer_main.py
# https://github.com/facebookresearch/vissl/blob/master/vissl/trainer/trainer_main.py#L89

class SelfSupervisionTrainer(object):
    def __init__(
        self,
        cfg: AttrDict,
        dist_run_id: str,
        checkpoint_path: str = None,
        checkpoint_folder: str = None,
        hooks: List[ClassyHook] = None,
    ):
        .....
        self.local_rank, self.distributed_rank = get_machine_local_and_dist_rank()
        self.task.build_event_storage_writers()
        
# Create the initiaite_vissl_event_storage() in train_task.py
# https://github.com/facebookresearch/vissl/blob/master/vissl/trainer/train_task.py
@register_task("self_supervision_task")
class SelfSupervisionTask(ClassificationTask):
    def __init__(self, config: AttrDict):
        super().__init__()
        .....
        self.event_storage_writers = []
        ....
    
    def build_event_storage_writers(self):
        from vissl.utils.events import JsonWriter
        
        self.event_storage_writers = [
           JsonWriter(f"{self.checkpoint_folder}/stdout.json")
        ]


# in the trainer, update the storage iteration
# https://github.com/facebookresearch/vissl/blob/master/vissl/trainer/trainer_main.py#L174
task.event_storage.iter = iteration

# In the hooks/log_hooks.py, use the writer to log to json
# https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L166-L206
for writer in task.event_storage_writers:
    writer.write()
```

### PR DESCRIPTION
**Pull request by akainth015**
Resolves #11 

- [x] Define classes
- [x] Update `SelfSupervisionTrainer` and `SelfSupervisionTask`
- [x] Add metrics from `log_hooks.py`
- [x] Implement and use the `JsonEventWriter`

Co-authored-by: Grace Omotoso

### CODE REVIEW

**Comment by akainth015** [comment]
Example `stdout.json` from `python3 -m unittest test_tasks.py`
```
{"batch_time": 1132.0, "epoch": 0.0, "eta": 1.1322011947631836, "iteration": 0, "loss": 3.00722, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 1014.0, "epoch": 0.0, "eta": 2.0299949645996094, "iteration": 0, "loss": 6.38364, "lr": 0.00047, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 1936.0, "epoch": 0.0, "eta": 1.9367702007293701, "iteration": 0, "loss": 1.3317, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 526.0, "epoch": 0.0, "eta": 0.526491641998291, "iteration": 0, "loss": 1.13592, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
```

**Comment by prigoyal** [comment]
thank you @akainth015 and @grace-omotoso  :) In the stdout.json , it looks like "iteration" is not getting updated. We should debug this. :)

File: vissl/trainer/trainer_main.py [hunk 1]
```diff
@@ -185,7 +187,7 @@ def train(self):
                         logging.info("CUDA cache cleared")
                     task = train_step_fn(task)
                     iteration_num += 1
```
Comment (prigoyal) [inline vissl/trainer/trainer_main.py:189]: we can update the iteration property of the storage from line 175 above in this module :)

Comment (akainth015) [reply vissl/trainer/trainer_main.py:189]: I intended update it here on line 190.
```
task.local_iteration_num = task.event_storage.iter = iteration_num
```

I think iteration being the same on all lines is the expected behavior for `tests/test_tasks.py`, but I noticed that [in this PR](https://github.com/MLH-Fellowship/vissl/pull/8#issuecomment-787131794) we had the following output where the iteration is `1` and not `0`.
```
{"batch_time": 1132.0, "epoch": 0.0, "eta": 1.1322011947631836, "iteration": 0, "loss": 3.00722, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 1014.0, "epoch": 0.0, "eta": 2.0299949645996094, "iteration": 0, "loss": 6.38364, "lr": 0.00047, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 1936.0, "epoch": 0.0, "eta": 1.9367702007293701, "iteration": 0, "loss": 1.3317, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 526.0, "epoch": 0.0, "eta": 0.526491641998291, "iteration": 0, "loss": 1.13592, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
```

Is there a different/additional place in the code where we should update the iteration number?

File: vissl/trainer/trainer_main.py [hunk 2]
```diff
@@ -185,7 +187,7 @@ def train(self):
...
-                    task.local_iteration_num = iteration_num
+                    task.local_iteration_num = task.event_storage.iter = iteration_num
```
Comment (prigoyal) [inline vissl/trainer/trainer_main.py]: reverting this + additionally set`task.event_storage.iter = iteration_num`

File: vissl/utils/events.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        """
+        return self._latest_scalars
+
+    @property
+    def iter(self):
+        """
+        Returns:
+            int: The current iteration number.
+        """
+        return self._iter
```
Comment (prigoyal) [inline vissl/utils/events.py:101]: insert pdb or print(self._iter)

File: vissl/utils/events.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        """
+        self._file_handle = PathManager.open(json_file, "a")
+
+    def write(self):
+        storage: VisslEventStorage = get_event_storage()
+        to_save = defaultdict(dict)
+
+        for k, (v, iter) in storage.latest().items():
+            # keep scalars that have not been written
+            to_save[iter][k] = v
```
Comment (prigoyal) [inline vissl/utils/events.py:136]: verify `print(to_save)` verify that the iteration value is `self._iter`

**Review by prigoyal — COMMENTED** [review]
-> remove any .json files -> `./checkpoints/my_output.json`
-> run a test to verify "epoch" gets updated as well - @prigoyal to share the command

File: vissl/utils/events.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    def clear_images(self):
+        return NotImplementedError
+
+    def clear_histograms(self):
+        return NotImplementedError
+
+    def put_histogram(self, hist_name, hist_tensor, bins=1000):
+        return NotImplementedError
+
+    def put_image(self, img_name, img_tensor):
```
Comment (prigoyal) [inline vissl/utils/events.py:116]: `3 x 224 x 224` -> `3 x H x W` -> RGB format image

**Comment by prigoyal** [comment]
@akainth015 , @grace-omotoso , sharing the test command for the "epoch" testing (as discussed above and in meeting today):
`python tools/run_distributed_engines.py config=test/cpu_test/test_cpu_resnet_simclr config.DATA.TRAIN.DATA_LIMIT=100 config.OPTIMIZER.num_epochs=2 config.CHECKPOINT.DIR="./checkpoints"`

File: vissl/utils/events.py (hunk 3 above)
Comment (grace-omotoso) [reply vissl/utils/events.py:101]: @prigoyal: We debugged the inconsistencies with the iteration value using the test you provided us: Please, find the log below:

```{"batch_time": 6273.0, "epoch": 0.0, "eta": 150.56117820739746, "iteration": 1, "loss": 1.3317, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5147.0, "epoch": 0.0, "eta": 102.94003486633301, "iteration": 4, "loss": 1.28832, "lr": 0.0022, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5190.0, "epoch": 0.0, "eta": 77.85301351547241, "iteration": 9, "loss": 1.24261, "lr": 0.00167, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5131.0, "epoch": 0.0, "eta": 51.3175384203593, "iteration": 14, "loss": 1.31571, "lr": 0.00095, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5178.0, "epoch": 0.0, "eta": 25.892525255680084, "iteration": 19, "loss": 0.75789, "lr": 0.00032, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5142.0, "epoch": 0.0, "eta": 0.0, "iteration": 24, "loss": 1.28264, "lr": 1e-05, "peak_mem_used": -1.0, "rank": 0.0}

Epoch is not being updated though.

File: vissl/hooks/log_hooks.py [hunk 6]
```diff
@@ -201,14 +201,11 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
...
-                    log_data[f"btime({self.btime_freq}iters) (ms)"] = rolling_btime
-                    log_data["rolling_eta"] = rolling_eta_str
-                logging.info(json.dumps(log_data))
-                with PathManager.open(
-                    f"{task.checkpoint_folder}/stdout.json", "a"
-                ) as fopen:
-                    fopen.write(json.dumps(log_data) + "\n")
-                    fopen.flush()
+                    evt_stg.put_scalars(
+                        rolling_btime=rolling_btime, rolling_eta=rolling_eta_str
```
Comment (akainth015) [inline vissl/hooks/log_hooks.py]: The `HistoryBuffer` can only store scalars, so we should only attempt to store a scalar
```suggestion
                        rolling_btime=rolling_btime, rolling_eta=rolling_eta_secs
```

File: vissl/utils/events.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+_VISSL_EVENT_STORAGE_STACK = []
+
+
+def get_event_storage():
+    return _VISSL_EVENT_STORAGE_STACK[-1]
+
+
+def create_event_storage():
+    _VISSL_EVENT_STORAGE_STACK.append(VisslEventStorage(1))
```
Comment (akainth015) [inline vissl/utils/events.py:16]: @prigoyal we put in a `1` here for starting iteration, but I think that may not work if we are resuming from a checkpoint. I think adding an optional `start_iter=0` parameter would be a good change for `create_event_storage`, but I'm not quite sure yet where I would get that from. Any pointers?

File: vissl/hooks/log_hooks.py (hunk 6 above)
Comment (prigoyal) [reply vissl/hooks/log_hooks.py]: yep! only attempt to store scalars.

File: vissl/utils/events.py (hunk 7 above)
Comment (prigoyal) [reply vissl/utils/events.py:16]: https://github.com/MLH-Fellowship/vissl/blob/master/vissl/trainer/trainer_main.py#L174 here the `task.iteration` contains the correct start_iter (whether start from scratch or resume from a checkpoint)

Comment (grace-omotoso) [reply vissl/utils/events.py:16]: @prigoyal We are still getting an iteration value of 0 at that point. Do you think we can hop on a call any time today?

Comment (prigoyal) [reply vissl/utils/events.py:16]: sorry @grace-omotoso , I was on PTO but we have our scheduled call today, let's use the time then :)

File: vissl/utils/events.py (hunk 3 above)
Comment (prigoyal) [reply vissl/utils/events.py:101]: can you share the log.txt file ? how many iterations ran  and does log.txt show updated epochs number? also where we are logging the epochs in this change, can you insert pdb and verify just like we did debugging for "iteration" ?

File: vissl/hooks/log_hooks.py [hunk 8]
```diff
@@ -173,42 +173,35 @@ def on_update(self, task: "tasks.ClassyTask") -> None:
...
-                    "iter": iteration,
-                    "lr": lr_val,
-                    "loss": loss_val,
-                    "btime(ms)": batch_time,
-                    "eta": eta_string,
-                    "peak_mem(M)": peak_mem_used,
-                }
+                evt_stg.put_scalars(
+                    rank=rank,
+                    epoch=train_phase_idx,
```
Comment (prigoyal) [inline vissl/hooks/log_hooks.py:184]: can you insert pdb here to verify the value of `train_phase_idx` when it changes and if this is properly logged to storage too?

**Review by prigoyal — APPROVED** [review]
merging this as we verified the output in the meeting yesterday! 

great work @akainth015 and @grace-omotoso :)
